File: 2nd_quarter/week01/Whaeun/뒤에 있는 큰 수 찾기.cpp
#include <string>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

vector<int> solution(vector<int> numbers) {
    vector<int> answer;
    stack<int> num;
    
    for(int i = numbers.size() - 1; i >= 0; i--){
        
        while(!num.empty() && numbers[i] >= num.top()){
            num.pop();
        }
        
        // no bigger num
        if(num.empty()){
            answer.push_back(-1);
        }
        else{
            answer.push_back(num.top());
        }
        
        num.push(numbers[i]);
    }
    
    reverse(answer.begin(), answer.end());
    
    return answer;
}

File: 2nd_quarter/week01/pepe/뒤에있는큰수찾기.py
"""
풀이 시간: 16:47 ~

n = 1,000,000 이므로 n^2은 안 됨
뒷 큰 수: 현재 값보다 뒤에 있고 && 큰 숫자

포인터 생성

numbers를 탐색하며, 포인터를 오른쪽으로 이동
이동한 포인터가 현재 값보다 크다면: answer에 추가
작다면: 오른쪽으로 이동
"""


def solution(numbers):
    stack = []
    answer = [-1] * len(numbers)

    for i in range(len(numbers)):
            while stack and numbers[stack[-1]] < numbers[i]:
                answer[stack.pop()] = numbers[i]
            stack.append(i)
    
    return answer

numbers = [9, 1, 5, 3, 6, 2]	
print(solution(numbers)) 